frontend/utils/distribution/group-income-distribution.js

```diff
@@ -1,40 +1,21 @@
 import { saferFloat } from '~/frontend/views/utils/currencies.js'
-import incomeDistribution from '~/frontend/utils/distribution/mincome-proportional.js'
+import parseMonthlyDistributionFromEvents from '~/frontend/utils/distribution/monthly-event-parser.js'
+import minimizeTotalPaymentsCount from '~/frontend/utils/distribution/payments-minimizer.js'
 
-export default function groupIncomeDistribution ({ state, getters, monthstamp, adjusted }: Object): any {
-  // the monthstamp will always be for the current month. the alternative
-  // is to allow the re-generation of the distribution for previous months,
-  // but that approach requires also storing the historical mincomeAmount
-  // and historical groupProfiles. Since together these change across multiple
-  // locations in the code, it involves less 'code smell' to do it this way.
-  // see historical/group.js for the ugly way of doing it.
-  const mincomeAmount = getters.groupMincomeAmount
-  const groupProfiles = getters.groupProfiles
-  const currentIncomeDistribution = []
-  for (const username in groupProfiles) {
-    const profile = groupProfiles[username]
-    const incomeDetailsType = profile && profile.incomeDetailsType
-    if (incomeDetailsType) {
-      const adjustment = incomeDetailsType === 'incomeAmount' ? 0 : mincomeAmount
-      const amount = adjustment + profile[incomeDetailsType]
-      currentIncomeDistribution.push({
-        name: username,
-        amount: saferFloat(amount)
-      })
-    }
-  }
-  let dist = incomeDistribution(currentIncomeDistribution, mincomeAmount)
-  if (adjusted) {
-    // if this user has already made some payments to other users this
-    // month, we need to take that into account and adjust the distribution.
-    // this will be used by the Payments page to tell how much still
-    // needs to be paid (if it was a partial payment).
+export default function groupIncomeDistribution (distributionEvents: Array<Object>, { mincomeAmount, adjusted = false, minimizeTxns = false, monthlyRated = true }: Object): any {
+  let dist = parseMonthlyDistributionFromEvents(distributionEvents, mincomeAmount, monthlyRated)
```

Comment:
So the lines 7 - 19 can be removed, and the `adjusted` parameter can be passed forward to the `parseMonthlyDistributionFromEvents` function. E.g. something along these lines:

```js
function distribution (events, options = { adjusted: false, minimized: false }) {
  dist = unadjustedDistributionFromEvents(events)
  if (options.adjusted) {
    payments = events.filter((evt) => evt.type === 'payment')
    for (const payment of payments) {
      for (const todo of dist) {
        if (todo.from === payment.from) {
          dist.amount = Math.max(0, dist.amount - payment.amount)
        }
      }
    }
  }
  if (options.minimized) {
    dist = minimize(dist)
  }
  return dist
}
unadjustedDistributionFromEvents = unadjustedDistributionFromEventsObjectVersion
function unadjustedDistributionFromEventsObjectVersion (events) {
  const users = {}
  for (const { data } of events.filter(evt => evt.type === 'haveNeed')) {
    users[data.who] = data.haveNeed
  }
  return mincomeProportional(users)
}

// Note: you can either pass in `mincome` as a parameter
// or, you can pass in `mincome` as an event in the `events` and extract it
function unadjustedDistributionFromEventsArrayVersion (events /* , mincome (?) */) {
  const users = []
  for (const { data } of events.filter(evt => evt.type === 'haveNeed')) {
    users.removeWhere(u => u.name === data.who)
    users.push({ name: data.who, haveNeed: data.haveNeed })
  }
  return mincomeProportional(users, mincome)
}
```

Reply:
As an alternative to passing in the `adjusted` to `parseMonthlyDistributionFromEvents`, you can do the check on that parameter here in this function. E.g. something like this:

```js
export default function groupIncomeDistribution (distributionEvents: Array<Object>, options: Object): any {
  const dist = (adjusted ? unadjustedDistribution : adjustedDistribution)(distributionEvents, options)
  return minimizeTxns ? minimizeTotalPaymentsCount(dist) : dist
}
```

Reply:
@taoeffect I am going to push my fix for cypress first. This is next on the TODO list.

Reply:
Correct me if I am wrong, but I think this is outdated because the `forgiveWithoutForget` replaces much of this code.

Marking resolved, for now.